### PR TITLE
negotiate the best static asset using supported encodings

### PR DESCRIPTION
**Pull request by mmerickel**
fixes https://github.com/Pylons/pyramid/issues/3432

Basically this adds support for a `content_encodings` argument in `add_static_view` and `static_view`. If supplied, then `static_view` will use the `accept-encoding` HTTP header to try and serve the appropriately-encoded file (preferring smaller files). This will also set the ``Vary`` header appropriately for files with multiple encodings.

The list of encodings comes from `mimetypes.encodings_map` and so it depends on the system. Modifying the map can be done to add support for other encodings like brotli.

```python
mimetypes.encodings_map['.br'] = 'br'
```

The locations of the files on disk are also cached by default unless `pyramid.reload_assets` is true.

### CODE REVIEW

File: src/pyramid/static.py [hunk 1]
```diff
@@ -94,46 +143,128 @@ def __call__(self, context, request):
         if path is None:
             raise HTTPNotFound('Out of bounds: %s' % request.url)
 
+        # normalize asset spec or fs path into resource_path
         if self.package_name:  # package resource
             resource_path = '%s/%s' % (self.docroot.rstrip('/'), path)
             if resource_isdir(self.package_name, resource_path):
                 if not request.path_url.endswith('/'):
-                    self.add_slash_redirect(request)
+                    raise self.add_slash_redirect(request)
```
Comment (bertjwregeer) [inline src/pyramid/static.py]: This changes the behaviour from previous versions of Pyramid, will this break anything?

**Comment by bertjwregeer** [comment]
I say yes... but Travis says no.

File: src/pyramid/static.py (hunk 1 above)
Comment (mmerickel) [reply src/pyramid/static.py]: It's not different, I just moved the raise outside the method instead of inside to be more explicit.

**Comment by mmerickel** [comment]
Thanks, yeah I'm writing up some narrative docs right now and changelog. I also obviously forgot to blackify.

File: src/pyramid/static.py (hunk 1 above)
Comment (bertjwregeer) [reply src/pyramid/static.py]: Oh, I see now. I missed it. :+1:

File: src/pyramid/static.py [hunk 2]
```diff
@@ -94,46 +143,126 @@ def __call__(self, context, request):
...
+        # only cache the results if reload is disabled
+        if not self.reload:
+            self.filemap[resource_name] = result
+        return result
+
+    def find_best_match(self, request, files):
+        """ Return ``(path | None, encoding)``."""
+        # if the client did not specify encodings then assume only the
+        # identity is acceptable
+        if not request.accept_encoding:
```
Comment (mmerickel) [inline src/pyramid/static.py]: short circuit `if len(files) < 2` here and return the identity path as well - no sense in doing the rest of the accept-header negotiation if we only have one possible result

**Pull request merged**